### DISCUSSION TITLE
Question about Cameras, Events and Engines...

### DISCUSSION THREAD

**Discussion by harzner** [coin3d/coin#537] [Q&A]
Hi peeps, 

I have a question about the interaction of Cameras, Events, and Engines...

Let's say that I have two objects in the scene graph right next to each other, and the Camera is looking at the left object. 

What I'm looking for is a solution for a camera-animation to "slide" from the left object to the right, triggered by a click on the RIGHT_ARROW key.

I am struggling to find a way to make a Camera move "swiftly" between stationary objects... triggered by a KEY

So far I understand that a camera can get a position without a translation node, it can look at things, but what I just can't get to work is a swift movement of the camera over time, in relation to some pretty stationary objects.

Any hints? I would be grateful!

I tried to use a SoOneShot Engine to be triggered by the key-event, then use the "ramp" to send that to an SoInterpolation, between the cameras current position, and its desitination, but that didn't work out.

I'm honestly quite lost... any ideas?

Thanks, and kind regards,
Harzner

**Comment by veelo** [comment]
I think your approach is not bad, but without seeing the code it is difficult to say why it doesn't work. When I implemented similar functionality long ago I decided to use Qt for the animation, based on QAbstractAnimation, QEasingCurve and QStateMachine. Then use [SbRotation::slerp()](https://www.coin3d.org/coin/classSbRotation.html#a04650493010ed066c2d055596f38c6f4) to interpolate orientation, and I interpolated the position manually. Not sure why I didn't use SoInterpolate. Edit: I think the reason for not using SoInterpolate was that if you want to rotate the camera around an object you want the camera to travel over a circular arc, not a straight line.

**Reply by harzner** [reply]
Hello Bastiaan, thank you so much for your reply!

My code is currently such a mess that it would be a waste of your time, but I think, I can now explain my problem much clearer:

Say that I have a little SoWinExaminerViewer, and a scene graph with my own Camera in it (that works fine), and some objects. 

I also have a callback function that is called when the user clicks on a key.

Currently, I am changing the position of the camera to the right, when the user clicks the right-arrow, and to the left, when the user clicks the left-arrow. Here's a snippet:

```
		if (SO_KEY_PRESS_EVENT(event, LEFT_ARROW)) {
			std::cout << "LEFT\r\n";
			initialPos.getValue(x, y, z);
			camera->position.setValue(x - 400.0, y, z);
			keyEvent->setHandled();
			return;
		}

		if (SO_KEY_PRESS_EVENT(event, RIGHT_ARROW)) {
			std::cout << "RIGHT\r\n";
			initialPos.getValue(x, y, z);
			camera->position.setValue(x + 400.0, y, z);
			keyEvent->setHandled();
			return;
		}

```

In my code, the position of the camera is changed instantly, but I would like the camera to "slide".

So, I need to somehow "update" the position of the camera repeatedly for a certain amount of time.

So, when the user clicks right-arrow, I guess I need some form of timer, and a callback function to update the camera's position until it reaches its destination. It seems that this "callback" is what I don't quite unterstand, yet.

I've seen some code in the ExaminerViewer that uses Sensors to keep an animation going, for example when you push an object into rotation with the mouse, but it's difficult to dig through all the code to put the pieces of the puzzle together.

Anyway, I'm onto reading about "lerp" functions, and Sensors, and it seems that I'm onto something, but thanks again for your reply!

Kind regards, 
Harzner

**Comment by harzner** [comment]
I think, I have found, what I need: An SoTimerSensor.

It sais in the description: "commonly used to trigger animation updates at a constant frame rate"

Pretty much exactly what I was looking for.

**Reply by harzner** [reply]
By the way: It works. The code is crude, and the motion not as "smooth" as I would like, but the funtionality I was looking for works.

Using an SoTimerSensor, I set up a callback routine for the camera-movement. 

When I click on the left-arrow button, my callback routine for key-events is called, in that function I activate the SoTimerSensor with sensor->schedule(), and from there, the callback for the sensor is called every 1/30 of a second.

In there I can now calculate the new position for the camera, and finally "unschedule" the sensor when the camera is done moving.

**Comment by VolkerEnderlein** [comment]
@harzner You may have a look onto `So@Gui@Viewer::seekToPoint(const SbVec3f & scenepos)` method implementation of SoGuiViewer component for SoQt or SoWin. It should give you some insight how an animated motion is done in the default viewers. It's the flashlight icon in the viewers GUI.

**Reply by harzner** [reply]
Hey Volker, 

thanks! I will do that. However, for the moment, I'm happy with the progress. I'm just a bit smarter now, than I was this morning 8)

Kind regards,
Harzner
